### ISSUE TITLE
Callback form of should chained off cy.window receives null when retried

### ISSUE THREAD

**Issue by david-sabata** [cypress-io/cypress#22587]
### Current behavior

My goal is to have a retriable check of a `window` property that my app exposes. To achieve that, I'm using [the callback form](https://docs.cypress.io/guides/core-concepts/retry-ability#Use-should-with-a-callback) of `should`.

The window object is correctly passed into the callback only for the first time. If the assertions fail and trigger a retry, the callback argument is `null`. (See the repro code)

Maybe I missed something, but I went through the docs and I don't think this is expected behavior.

<img width="418" alt="image" src="https://user-images.githubusercontent.com/305251/176432700-497b3a45-0513-4dbb-b8ee-7a3157bc9037.png">

### Desired behavior

The callback of `should` always receives the `window` object

### Test code to reproduce

```js
it('retries should on window object', () => {
    cy.window().should(win => {
      // this logs <window> on the first run, but null on the next
      cy.log('window =', win).then(() => {
        // failing to trigger a retry
        expect(true).to.equal(false)
      })
    })
})
```

### Cypress Version

10.3.0

**Comment by david-sabata** [comment]
Possibly related to #8346 ?

**Comment by emilyrohrbough** [comment]
@david-sabata The issue you have logged is indeed a duplicate of #8346 (thank you for linking!) The `cy.log()` has a bug because it logs twice.

With the example you linked out, because the `.then()` command fails, it won't actually trigger a retry.

I assume your example is a rough high-level example of the behavior you are trying to accomplish? For the `cy.window()` command to re-poll when the `.should()` command fails, you'll want to stick with sync assertions and avoid non-cypress commands.

**Comment by BlueWinds** [comment]
Very interesting, thanks for the precise repro. I'm currently working in this area - how subjects are determined and how commands are retried - in relation to https://github.com/cypress-io/cypress/issues/7306.

The short version is that `should` is not really designed for this use case, and we don't provide anything that *is* designed for this use case either. ^^;; I'm aiming to address that in my work. I'm going to assign myself here so I remember to follow up on this specific case.

A temporary workaround would be to ensure that you don't have any cypress commands inside your `should` callback:

```
it.only('retries should on window object', () => {
      let count = 0
      cy.window().should(win => {
        count++
        if (count < 4) { throw new Error('foobar') }
        expect(win).to.haveOwnProperty('location')
      })
    })
```

This passes; it's only when the should callback contains other cypress commands that cy gets confused over what the subject should be during retries.

**Comment by BlueWinds** [comment]
For anyone following along, this will be "fixed" in 11.0.0. Fixed is in quotes because it'll be fixed by completely disallowing Cypress commands from inside `.should()`, throwing an explicit error that this is unsupported rather than doing weird things (like logging twice, losing track of the subject, etc).

**Comment by mschile** [comment]
The code for this is done in https://github.com/cypress-io/cypress/pull/23734, but has yet to be released.
We'll update this issue and reference the [changelog](https://on.cypress.io/changelog) when it's released.

**Comment by emilyrohrbough** [comment]
~This was released with 11.0.0.~
EDIT: This was NOT released with v11.0.0. This will be released in v12.0.0.
